File: src/entities/Promotion.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  OneToMany,
  JoinTable,
  Unique,
  RelationId,
} from 'typeorm';
import { Product } from './Product';
import { User } from './User';
import { Vote } from './Vote';
import { Brand } from './Brand';
import { Store } from './Store';
import { MesuringUnits } from '@entities/utils';

export enum PromotionType {
  SIMPLE = 0,
  TWOSECONDHALF = 1,
  THREEFORTWO = 2,
  UNDEFINE = 3,
}

@Entity()
@Unique(['promotion', 'beginDate', 'endDate', 'type', 'product', 'storeId'])
@Unique(['promotion', 'beginDate', 'endDate', 'type', 'product', 'brandId'])
export class Promotion {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  description: string;

  @Column('decimal', { precision: 6, scale: 2 })
  price: number;

  @Column('decimal', { precision: 6, scale: 2 })
  promotion: number;

  @Column()
  beginDate: Date;

  @Column({ nullable: true })
  endDate: Date;

  @Column('enum', { enum: PromotionType, default: PromotionType.SIMPLE })
  type: number;

  @Column({ default: 1 })
  quantity: number;

  @Column('enum', { enum: MesuringUnits, default: MesuringUnits.UNIT })
  unit: number;

  @Column({ nullable: true })
  userId: number;

  @Column({ nullable: true })
  storeId: number;

  @Column({ nullable: true })
  brandId: number;

  @ManyToOne(type => Product, product => product.barcode, {
    onDelete: 'CASCADE',
  })
  product: Product;

  @ManyToOne(type => User, { nullable: true, onDelete: 'CASCADE' })
  user: User;

  @ManyToOne(type => Store, { nullable: true, onDelete: 'CASCADE' })
  store: Store;

  @ManyToOne(type => Brand, { nullable: true, onDelete: 'CASCADE' })
  brand: Brand;

  @OneToMany(type => Vote, vote => vote.promotion)
  @JoinTable()
  votes: Vote[];
}

File: scrappers/models/Promotion.py
from scripts.utils import brand_by_name_get, product_by_name_get


class Promotion:
    def __init__(self, store_name, begin_date, end_date, name, price, promo):
        self.store_name = store_name
        self.beginDate = begin_date
        self.endDate = end_date
        self.name = name
        self.price = float(price.replace(",", ".")) if "," in price else float(price)
        self.promotion = float(promo.replace(",", ".")) if "," in promo else float(promo)

    def __str__(self):
        return f'Store: {self.store_name} / Product name: {self.name} / Product price: {self.price} / Promotion: {self.promotion}'

    def convertToPromotion(self, url, user_id):
        brand_id = brand_by_name_get(url + "brands/" + self.store_name)
        if brand_id is not None:
            brand_id = brand_id["id"]
        product = product_by_name_get(url + "products/name/" + self.name)
        if product is not None:
            product = product["barcode"]
        else:
            return None
        return MetaPromotion("", self.beginDate, self.endDate, user_id, None, brand_id, product, self.price, self.promotion)


class MetaPromotion:
    def __init__(self, description, begin_date, end_date, user_id, store_id, brand_id, product, price, promotion):
        self.description = description
        self.beginDate = begin_date
        self.endDate = end_date
        self.userId = user_id
        self.storeId = store_id
        self.brandId = brand_id
        self.product = product
        self.price = price
        self.promotion = promotion
